src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets

```
@@ -55,20 +55,20 @@
     <Error Condition="'$(PublishAot)' == 'true' and '$(PublishSingleFile)' == 'true'" Text="PublishAot and PublishSingleFile cannot be specified at the same time." />
 
     <!-- Fail with descriptive error message for common unsupported cases. -->
-    <Error Condition="'$(DisableUnsupportedError)' != 'true' and '$(OS)' == 'Windows_NT' and !$(RuntimeIdentifier.StartsWith('win'))"
+    <Error Condition="'$(DisableUnsupportedError)' != 'true' and '$(OS)' == 'Windows_NT' and '$(_targetOS)' != 'win'"
       Text="Cross-OS native compilation is not supported." />
 
-    <Error Condition="'$(DisableUnsupportedError)' != 'true' and '$(OS)' != 'Windows_NT' and $(RuntimeIdentifier.StartsWith('win'))"
+    <Error Condition="'$(DisableUnsupportedError)' != 'true' and '$(OS)' != 'Windows_NT' and '$(_targetOS)' == 'win'"
       Text="Cross-OS native compilation is not supported." />
 
-    <Error Condition="'$(DisableUnsupportedError)' != 'true' and !($(RuntimeIdentifier.EndsWith('x64')) or $(RuntimeIdentifier.EndsWith('arm64')))"
+    <Error Condition="'$(DisableUnsupportedError)' != 'true' and '$(_targetArchitecture)' != 'x64' and '$(_targetArchitecture)' != 'arm64'"
       Text="Native compilation does not support targeting $(RuntimeIdentifier) yet." />
 
-    <Error Condition="'$(DisableUnsupportedError)' != 'true' and !('$(IlcHostArch)' == 'x64' or '$(IlcHostArch)' == 'arm64')"
+    <Error Condition="'$(DisableUnsupportedError)' != 'true' and !('$(_hostArchitecture)' == 'x64' or '$(_hostArchitecture)' == 'arm64')"
       Text="Native compilation can run on x64 and arm64 hosts only." />
 
-    <Error Condition="'$(IlcHostPackagePath)' == '' and '$(RuntimePackagePath)' != '' and ('$(IlcHostArch)' == 'x64' or '$(IlcHostArch)' == 'arm64')"
-      Text="Add a PackageReference for '$(IlcHostPackageName)' to allow cross-compilation for $(_targetArchitecture)" />
+    <Error Condition="'$(IlcHostPackagePath)' == '' and '$(RuntimePackagePath)' != '' and ('$(_hostArchitecture)' == 'x64' or '$(_hostArchitecture)' == 'arm64')"
+      Text="Add a PackageReference for '$(_hostPackageName)' to allow cross-compilation for $(_targetArchitecture)" />
```

I think this would be far more useful check:

```suggestion
    <Error Condition="'$(IlcHostPackagePath)' == ''"
      Text="ILCompiler host package '$(_hostPackageName)' was not resolved or restored correctly" />
    <Error Condition="'$(RuntimePackagePath)' == ''"
      Text="ILCompiler target package '$(_targetPackageName)' was not resolved or restored correctly" />
```

The original check is not relevant anymore since the SDK is supposed to resolve and restore both the host and target packages. There are certainly possible issues with the logic but the check no longer captures them correctly.

I'm not sure about the wording of the error. We could suggest to add `PackageReference` but it's really just a workaround for SDK bug.

---

> The original check is not relevant anymore

Yes, most of these checks predate SDK integration. I think we should cut another PR for rationalizing these error post-integration. These fixes may seem simple, but they require a lot of manual testing on entire matrix to ensure user experience is not broken; because CI coverage for PublishAot is falling short in some cases: https://github.com/dotnet/runtime/pull/82590#issuecomment-1444356773.

---

Agreed, let's not block this PR on it.